SVGRenderStyle should repaint on resolved color change

mdubet opened the pull request:
#### f5b520fdec0edc2996ac754c2fa8f13fd2a7fb05
<pre>
SVGRenderStyle should repaint on resolved color change
<a href="https://bugs.webkit.org/show_bug.cgi?id=250718">https://bugs.webkit.org/show_bug.cgi?id=250718</a>
rdar://102904403

Reviewed by Antti Koivisto.

Before this patch, the diff algorithm would naively
compare the color properties value without taking into
account the dynamic nature of &quot;currentcolor&quot;.

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/svg-currentcolor-dynamic-inherit-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/svg-currentcolor-dynamic-inherit.html: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayout const):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::diff const):
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
(WebCore::colorChangeRequiresRepaint):
(WebCore::SVGRenderStyle::changeRequiresLayout const):
(WebCore::SVGRenderStyle::changeRequiresRepaint const):
(WebCore::SVGRenderStyle::diff const): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/259082@main">https://commits.webkit.org/259082@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8089e5e40a8583b8ae52edd73e879f2d37f461e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103796 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12913 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36739 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113021 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173336 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107745 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13936 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3808 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96044 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112136 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109567 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10750 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93810 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38466 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80117 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6270 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26805 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6438 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3341 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12422 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46328 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6249 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8206 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/css/svg-currentcolor-dynamic-inherit.html:
```diff
@@ -0,0 +1,34 @@
+</head>
+<body>
+
+<script src="/resources/testdriver.js"></script>
```
You can't use absolute URLs in tests in `fast`

mdubet replied:
Fixed

smfr commented on Source/WebCore/rendering/style/SVGRenderStyle.cpp:
```diff
@@ -154,72 +175,72 @@ StyleDifference SVGRenderStyle::diff(const SVGRenderStyle& other) const
+        || colorDiff(m_strokeData->visitedLinkPaintColor, other.m_strokeData->visitedLinkPaintColor, currentColorDiffers)
+        || m_strokeData->visitedLinkPaintUri != other.m_strokeData->visitedLinkPaintUri
+        || m_strokeData->visitedLinkPaintType != other.m_strokeData->visitedLinkPaintType)
+        return true;
```
Why does a change of paintColor require layout?

mdubet replied:
I doubt it does actually (like opacity which is explicitly move to the repaint part) but I don't want to mix several changes in this patch.

anttijk replied:
It is clearly wrong to provide currentColorDiffers bit to changeRequiresLayout. Current color change can't affect layout. You should try to remove the color change layout here.

anttijk commented on Source/WebCore/rendering/style/SVGRenderStyle.cpp:
```diff
@@ -137,15 +137,36 @@ void SVGRenderStyle::copyNonInheritedFrom(const SVGRenderStyle& other)
+    return StyleDifference::Equal;
+}
+
+static bool colorDiff(const StyleColor& a, const StyleColor& b, bool currentColorDiffers)
```
Something more descriptive like requiresRepaintForColorChange?

anttijk commented on Source/WebCore/rendering/style/SVGRenderStyle.cpp:
```diff
@@ -137,15 +137,36 @@ void SVGRenderStyle::copyNonInheritedFrom(const SVGRenderStyle& other)
+
+static bool colorDiff(const StyleColor& a, const StyleColor& b, bool currentColorDiffers)
+{
+    if (a != b)
```
I suppose if alpha stays 0 a color change wouldn't actually need a repaint.

mdubet replied:
Indeed I think there are ways to be a bit smarter on this function (and reuse it for RenderStyle attributes also). I will open a bugzilla for this.

anttijk commented on Source/WebCore/rendering/style/SVGRenderStyle.cpp:
```diff
@@ -137,15 +137,36 @@ void SVGRenderStyle::copyNonInheritedFrom(const SVGRenderStyle& other)
+        return StyleDifference::Repaint;
+
+    return StyleDifference::Equal;
+}
```
Does this still have clients? I thought you deleted the only one. If not this should be deleted?

mdubet replied:
No indeed, it was the only usage. Removed.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1278,22 +1278,16 @@ StyleDifference RenderStyle::diff(const RenderStyle& other, OptionSet<StyleDiffe
+        return StyleDifference::Layout;
 
     if (changeRequiresLayout(other, changedContextSensitiveProperties))
         return StyleDifference::Layout;
```
To be consistent with how other branches work you'd call `m_svgStyle->changeRequiresLayout`  from within `RenderStyle::changeRequiresLayout()`

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1278,22 +1278,16 @@ StyleDifference RenderStyle::diff(const RenderStyle& other, OptionSet<StyleDiffe
-        if (svgChange == StyleDifference::Layout)
-            return svgChange;
-    }
+    bool currentColorDiffers = m_inheritedData->color != other.m_inheritedData->color;
```
Or you could have just provided currentColorDiffers bit to m_svgStyle->diff to minimize the need to refactor. I'm not sure it is an actual improvement, returning StyleDifference is more flexible.

mdubet replied:
I think splitting the function in the same way in RenderStyle/SVGRenderStyle provides more flexibility (for example now, the repaint change from SVG only occurs after the RepaintLayer or the LayoutPositionedMovementOnly). This would be cumbersome to write with only the diff function.

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1278,22 +1278,16 @@ StyleDifference RenderStyle::diff(const RenderStyle& other, OptionSet<StyleDiffe
-    if (svgChange != StyleDifference::Equal)
-        return svgChange;
+    if (m_svgStyle->changeRequiresRepaint(other.m_svgStyle, currentColorDiffers))
+        return StyleDifference::Repaint;
```
Similarly with `changeRequiresRepaint`

anttijk commented on Source/WebCore/rendering/style/SVGRenderStyle.h:
```diff
@@ -44,6 +44,8 @@ class SVGRenderStyle : public RefCounted<SVGRenderStyle> {
 
     StyleDifference diff(const SVGRenderStyle&) const;
+    bool changeRequiresRepaint(const SVGRenderStyle& other, bool currentColorDiffers = false) const;
+    bool changeRequiresLayout(const SVGRenderStyle& other, bool currentColorDiffers = false) const;
```
You shouldn't provide default arguments here. You really can't compute changeRequiresRepaint correctly without one. There shouldn't currentColorDiffers argument for changeRequiresLayout.

mdubet committed:
SVGRenderStyle should repaint on resolved color change
https://bugs.webkit.org/show_bug.cgi?id=250718
rdar://102904403

Reviewed by Antti Koivisto.

Before this patch, the diff algorithm would naively
compare the color properties value without taking into
account the dynamic nature of "currentcolor".

* LayoutTests/imported/w3c/web-platform-tests/svg/painting/svg-currentcolor-dynamic-inherit-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/svg/painting/svg-currentcolor-dynamic-inherit.html: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayout const):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::diff const):
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
(WebCore::colorChangeRequiresRepaint):
(WebCore::SVGRenderStyle::changeRequiresLayout const):
(WebCore::SVGRenderStyle::changeRequiresRepaint const):
(WebCore::SVGRenderStyle::diff const): Deleted.
* Source/WebCore/rendering/style/SVGRenderStyle.h:

Canonical link: https://commits.webkit.org/259082@main
